[Merged by Bors] - feat(data/sym2) Defining the symmetric square (unordered pairs)

kmill opened the pull request:
This adds a type for the symmetric square of a type, which is the quotient of the cartesian square by permutations.  These are also known as unordered pairs.

Additionally, this provides some definitions and lemmas for equalities, functoriality, membership, and a relationship between symmetric relations and terms of the symmetric square.

I preferred `sym2` over `unordered_pairs` out of a combination of familiarity and brevity, but I could go either way for naming.


---
<!-- put comments you want to keep out of the PR commit here -->

kmill committed:
feat(data/sym2) Defining the symmetric square (unordered pairs)

This adds a type for the symmetric square of a type, which is the
quotient of the cartesian square by permutations.  These are also
known as unordered pairs.

We provide some definitions and lemmas for equalities, functoriality,
and a relationship between symmetric relations and terms of the
symmetric square.

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+attribute [refl] rel.refl
+
+@[symm] lemma rel.symm {x y : α × α} : rel α x y → rel α y x :=
+by { tidy, cases a, exact a, apply rel.swap }
```
please only use `tidy` to close a goal. The behavior of `tidy` can easily change in the future (for example new `simp` lemmas) and this can leave the proof state differently after you call tidy, which will break your proof. 

In this case, `intro a` probably suffices.

Please also do this with all occurrences of `tidy` below.

kim-em replied:
`tidy?` will report (somewhat verbosely) what it did.

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+lemma eq_swap {a b : α} : ⟦(a, b)⟧ = ⟦(b, a)⟧ :=
+by { rw quotient.eq, apply rel.swap }
+
+lemma other_unique (a b c : α) : ⟦(a, b)⟧ = ⟦(a, c)⟧ ↔ b = c :=
```
```suggestion
lemma congr_right (a b c : α) : ⟦(a, b)⟧ = ⟦(a, c)⟧ ↔ b = c :=
```

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+This is the type for the symmetric square (i.e., the type of unordered pairs).
+-/
+@[reducible]
+def sym2 (α : Type u) [h : setoid (α × α)] := quotient h
```
```suggestion
def sym2 (α : Type u) := quotient h
```
You want to use the relation you just defined, but you had it as variable.

kmill replied:
I was trying to make sure `quotient` was given the same instance that the double-bracket notation finds.  Would this be better?
```
def sym2 (α : Type u) := quotient (rel.setoid α)
```

fpvandoorn replied:
Oh yes, that's what I meant. The problem with your definition is that if I define a new instance of type `setoid (α × α)`, then `sym2` will magically change definitions (and all your proofs in the rest of the file will break).

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+A type `α` is naturally included in the diagonal of `α × α`, this function gives the image
+of this diagonal in `sym2 α`.
+-/
+def incl (x : α) : sym2 α := ⟦(x, x)⟧
```
I would use a slightly more descriptive name, like `diag` (or `single`).

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+/--
+Checks whether a given element of `sym2 α` is in the image of `incl`.
+-/
+def is_diag (z : sym2 α) : Prop :=
```
Is there a reason you want to state `is_diag z` instead of `z \in range incl`? 
You could always use the latter and not even define the former.
Then `diag_to_eq` becomes `(h : ⟦(x, y)⟧ \in range incl) : x = y`

kmill replied:
It looks like I don't use it anymore, so I'm going to delete it.  Good points.

kmill replied:
Actually, I ended up adding it back in but defined as
```
def is_diag (z : sym2 α) : Prop := z ∈ set.range (@diag α)
```
This is so I can have a reasonable `decidable_pred` instance, which I had needed at one point and anticipate needing again.

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+def image_incl_is_diag (z : sym2 α) (h : is_diag z) : ∃ x, z = incl x :=
+by tidy
+
+lemma diag_to_eq (x y : α) (h : is_diag ⟦(x, y)⟧) : x = y :=
```
please make this an `iff` statement, of the form `is_diag ⟦(x, y)⟧ \iff x = y`. `iff` statements are generally more useful than one implication. (and update its name appropriately)

fpvandoorn replied:
Lemmas that are about elements of `α × α` are a bit more general when formulated using `z : α × α` and then using `z` and `z.1` and `z.2` for its components.
Not every element in `α × α` is definitionally equal to a pair of two elements.

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+def is_diag (z : sym2 α) : Prop :=
+quotient.rec_on z (λ y, y.1 = y.2) (by { tidy, cases p; refl, cases p; refl })
+
+def image_incl_is_diag (z : sym2 α) (h : is_diag z) : ∃ x, z = incl x :=
```
please make this an `iff` statement (but it can be removed if you adopt the suggestion above, because then it's just a property of range)

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+/--
+The functor `sym2` is functorial, and this constructs induced maps.
+-/
+def induce {α β : Type u} (f : α → β) : sym2 α → sym2 β :=
```
this is usually called `map`. Al
```suggestion
def map {α β : Type*} (f : α → β) : sym2 α → sym2 β :=
```
This is usually called map. Also, you should not force `\a` and `\b` to live in the same universe

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+The functor `sym2` is functorial, and this constructs induced maps.
+-/
+def induce {α β : Type u} (f : α → β) : sym2 α → sym2 β :=
+λ z, quotient.rec_on z (λ x, ⟦(f x.1, f x.2)⟧)
```
Please use `quotient.map` and `prod.map`

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+end
+
+@[simp]
+def induce.id : sym2.induce (@id α) = id :=
```
I would call this `map_id` (there is no `induce`/`map` namespace you want to open or projection notation you can use)

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+def induce.id : sym2.induce (@id α) = id :=
+by tidy
+
+def induce.functorial {α β γ: Type u} {g : β → γ} {f : α → β} : sym2.induce (g ∘ f) = sym2.induce g ∘ sym2.induce f :=
```
`map_comp`

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+def induce.functorial {α β γ: Type u} {g : β → γ} {f : α → β} : sym2.induce (g ∘ f) = sym2.induce g ∘ sym2.induce f :=
+by tidy
+
+instance is_diag.decidable_pred (α : Type u) [decidable_eq α] : decidable_pred (@is_diag α) :=
```
Once `diag_to_eq` is an `iff` statement, that should make this proof very easy.

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+This is a predicate that determines whether a given term is a member of a term of the symmetric square.
+From this point of view, the symmetric square is the subtype of cardinality-two multisets on `α`.
+-/
+def mem (x : α) (z : sym2 α) : Prop :=
```
Make this an `has_mem` instance, so that you can write `x \in z`.
You could consider defining this as `z \in range (\la y, ⟦(x, y)⟧)`.

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+/--
+Given an element of a term of the symmetric square (using `vmem`), retrieve the other element.
+-/
+lemma other {a : α} {p : sym2 α} (h : vmem a p) : α := h.val
```
```suggestion
def vmem.other {a : α} {p : sym2 α} (h : vmem a p) : α := h.val
```
You can then refer to this element with `h.other`

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+The defining property of the other element is that it can be used to
+reconstruct the term of the symmetric square.
+-/
+lemma other_spec (a : α) (z : sym2 α) (h : vmem a z) : z = ⟦(a, other h)⟧ :=
```
```suggestion
lemma other_spec {a : α} {z : sym2 α} (h : vmem a z) : z = ⟦(a, h.other)⟧ :=
```

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+This is the `mem`-based version of `other`.
+-/
+noncomputable
+lemma mem_other {a : α} {z : sym2 α} (h : mem a z) : α :=
```
```suggestion
noncomputable def mem.other {a : α} {z : sym2 α} (h : mem a z) : α :=
```

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+lemma mem_other {a : α} {z : sym2 α} (h : mem a z) : α :=
+classical.some h
+
+lemma mem_other_spec (a : α) (z : sym2 α) (h : mem a z) : ⟦(a, mem_other h)⟧ = z :=
```
```suggestion
lemma mem_other_spec {a : α} {z : sym2 α} (h : mem a z) : ⟦(a, h.other)⟧ = z :=
```

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,201 @@
+  (x = z ∧ y = w) ∨ (x = w ∧ y = z) :=
+by { rw quotient.eq at h, cases h; tidy }
+
+lemma is_one_of {a b c : α} (h : mem a ⟦(b, c)⟧) : a = b ∨ a = c :=
```
```suggestion
lemma mem_iff {a b c : α} : mem a ⟦(b, c)⟧ ↔ a = b ∨ a = c :=
```

fpvandoorn reviewed (commented):
Thank you for your pull request!

Do you have a particular use case for this file already? Generally you get a much better library if you work towards a particular goal, and add things that you need (and lemmas similar to those) than if you just start writing a library without a goal in mind.

For example, are `mem` and `vmem` both useful?

kmill committed:
feat(data/sym2) Defining the symmetric square (unordered pairs)

Incorporated suggestions from code review.

kmill committed:
More documentation and irreflexive symmetric relation lemma

Added a mention that one of the key definitions is `sym2.from_rel`,
which transforms a symmetric relation into set membership.

Also added a lemma that is useful for simple graphs.

kmill committed:
Removed simp attribute

Also reformatted the lemma statement.

kmill commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,263 @@
+  h.other = mem.other h' :=
+by rw [←congr_right a, ←vmem_other_spec h, mem_other_spec]
+
+lemma eq_pairs {x y z w : α} :
```
Should I rename this `eq_iff` or something else?

jcommelin commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,263 @@
+end sym2
+
+/--
+This is the type for the symmetric square (i.e., the type of unordered pairs).
```
```suggestion
`sym2 α` is the symmetric square of `α`, in other words, the type of unordered pairs.
It is in a natural but noncomputable way equivalent to multisets of cardinality 2, see TODO.
```

jcommelin commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,263 @@
+This is also known as the type of unordered pairs.
+
+From the point of view that an unordered pair is equivalent to a
+multiset of cardinality two, we have `sym2.mem` (the `has_mem`
```
I would prefer having an `equiv` with `{s : multiset X // s.card = 2}`, and I suggest explicitly naming this result in the docstring for `sym2`, to make the connection with the multiset API.

kmill replied:
Defined it (in the newest commit).  It seemed easiest to define symmetric powers in general due to some issues I was having getting a `quot.rec` to work out properly...  Now that I got it all to work, I could probably extract a more direct proof, if you think that would be better.

jcommelin replied:
@kmill Thanks a lot, I think it's helpful. However, now the definition of `sym` looks a bit out of place. I suggest that you start a new file for this `data/sym` and put the definition (also `sym'`) and basic lemmas in there.
I won't ask you to flesh out the API of `sym` any further, just put a TODO/warning at the top of the file, that this is only there as supporting material for `sym2`.
Another PR could add a lot more stuff to the `sym` file.

kmill replied:
@jcommelin Ok, done!

kmill committed:
Added equivalence to cardinality-two multisets

To do this in a clean way, defined symmetric powers in general and an
equivalence between two potential definitions (subtype of multisets,
or quotient of vectors).  The difference is whether the
cardinality/length condition is outside or inside the quotient.

kmill committed:
Inhabited instances to make linter happy

kmill committed:
Fixed typo in documentation, simplified inhabited instance

kmill committed:
Moved sym definitions to data.sym

And reorganized some of data.sym2 accordingly.

jcommelin commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,358 @@
+  ⟦(x, y)⟧ = ⟦(z, w)⟧ ↔ (x = z ∧ y = w) ∨ (x = w ∧ y = z) :=
+by { split; intro h,
+     { rw quotient.eq at h, cases h; tidy, },
+     { cases h; rw [h.1, h.2], rw eq_swap, } }
```
```suggestion
begin
  split; intro h,
  { rw quotient.eq at h, cases h; tidy, },
  { cases h; rw [h.1, h.2], rw eq_swap, }
end
```

jcommelin commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,358 @@
+  simp,
+  simp,
+  split; apply sym,
+end))
```
```suggestion
λ z, quotient.rec_on z (λ z, r z.1 z.2)
begin
  intros z w p,
  cases p,
  simp,
  simp,
  split; apply sym,
end
```

jcommelin commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,358 @@
+    refl,
+    refl,
+  end
+}
```
```suggestion
  end }
```

jcommelin reviewed (commented):
Looks good to me! Some minor stylistic issues, but I think after that this is ready for merging.
@fpvandoorn Do you have further comments?

kmill committed:
Fixing stylistic issues

fpvandoorn commented on src/data/sym.lean:
```diff
@@ -0,0 +1,74 @@
+/--
+Multisets of cardinality n are equivalent to length-n vectors up to permutations.
+-/
+def sym_equiv_sym' (α : Type*) (n : ℕ) : equiv (sym α n) (sym' α n) :=
```
```suggestion
def sym_equiv_sym' (α : Type*) (n : ℕ) : sym α n ≃ sym' α n :=
```

fpvandoorn commented on src/data/sym2.lean:
```diff
@@ -0,0 +1,364 @@
+
+-- Instances to make the linter happy
+
+instance inhabited_vmem [inhabited α] : inhabited (vmem (default α) (diag (default α))) :=
```
Oh sorry, I didn't look carefully why the linter was complaining, I thought it was about the definition `sym2` itself (it isn't complaining about `sym2` because you marked it as reducible).

This instance is useless, and the linter shouldn't complain about `vmem`. Please remove this instance and add `@[nolint has_inhabited_instance]` in front of the definition of `vmem`.

fpvandoorn reviewed (commented):
Looking good, including all additions.

Feel free to merge it after incorporating my comments (or let someone else do it if you're unsure).

bors d+

kmill committed:
A style change and better appeasement of linter
